File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.py
from collections import deque
def solution(maps):
    answer = []
    N = len(maps)
    M = len(maps[0])
    visited = [[False] * M for _ in range(N)]
    
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]
    
    for i in range(N):
        for j in range(M):
            if maps[i][j] != 'X' and not visited[i][j]:
                q = deque()
                q.append((i, j))
                visited[i][j] = True
                tmp = 0
                while q:
                    r, c = q.popleft()
                    tmp += int(maps[r][c])
                    
                    for k in range(4):
                        nr = r + dr[k]
                        nc = c + dc[k]
                        if 0 <= nr < N and 0 <= nc < M:
                            if maps[nr][nc] != 'X' and not visited[nr][nc]:
                                q.append((nr, nc))
                                visited[nr][nc] = True
                answer.append(tmp)
    answer.sort()
    if not answer:
        answer = [-1]
    return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.ArrayList;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Collections;

class Solution {
    public int[] solution(String[] maps) {
        ArrayList<Integer> arr = new ArrayList<>();
        int[] answer = {};
        int N = maps.length;
        int M = maps[0].length();
        boolean[][] visited = new boolean[N][M];
        int[] dr = new int[] {-1, 1, 0, 0};
        int[] dc = new int[] {0, 0, -1, 1};
        
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (maps[i].charAt(j) != 'X' && !visited[i][j]) {
                    Queue<int[]> q = new LinkedList<>();
                    int tmp = 0;
                    q.offer(new int[] {i, j});
                    visited[i][j] = true;
                    
                    while(!q.isEmpty()) {
                        int[] cur = q.poll();
                        tmp += maps[cur[0]].charAt(cur[1]) - '0';
                        for (int k = 0; k < 4; k++) {
                            int nr = cur[0] + dr[k];
                            int nc = cur[1] + dc[k];
                            
                            if (0 <= nr && nr < N && 0 <= nc && nc < M) {
                                if (maps[nr].charAt(nc) != 'X' && !visited[nr][nc]) {
                                    q.offer(new int[] {nr, nc});
                                    visited[nr][nc] = true;
                                }
                            }
                        }
                    }
                    arr.add(tmp);
                } 
            }
        }
        if (arr.isEmpty()) {
            answer = new int[] {-1};
        } else {
            Collections.sort(arr);
            answer = new int[arr.size()];
            for (int i = 0; i < arr.size(); i++) {
                answer[i] = arr.get(i);
            }
        }
        
        return answer;
    }
}